File: src/pipe.cpp
//! \file pipe.cpp
#include "opensd/pipe.h"

#include <iostream>

#include "opensd/error.h"
#include "opensd/xml_interface.h"
#include "opensd/vector.h"

// #include <string>
// #include <cstdlib>  // for std::exit
// #include <dlfcn.h>  // for dynamic loading of libraries (Linux)
// #include <filesystem> // C++17 for current directory

namespace opensd {

//==============================================================================
// Global variables
//==============================================================================

//==============================================================================
// Wall implementation
//==============================================================================

Wall::Wall(double thk, const std::string& solname, const std::string& sollib, const std::string& restraint) {

  c1 = 5.0 / 4.0 - 0.26;

  this->thk = thk;
}

//==============================================================================
// Pipe implementation
//==============================================================================

Pipe::Pipe(pugi::xml_node pipe_node)
{
  if (check_for_node(pipe_node, "identifier")) {
    this->identifier = get_node_value(pipe_node, "identifier");
  } else {
    fatal_error("Must specify identifier of pipe in geometry XML file.");
  }

  this->diameter  = stod(get_node_value(pipe_node, "diameter"));
  this->length    = stod(get_node_value(pipe_node, "length"));
  this->dnode_str = get_node_value(pipe_node, "dnode");
  this->unode_str = get_node_value(pipe_node, "unode");
  this->ncell     = stod(get_node_value(pipe_node, "ncell"));
  this->cfarea    = stod(get_node_value(pipe_node, "cfarea"));
  this->unode = nullptr;
  this->dnode = nullptr;
  double ufrac;
  double dfrac;
  double delx;
  double delz;
  this->roughness = stod(get_node_value(pipe_node, "roughness"));
  double fricopt;

}

void Pipe::save_to_hdf5(hid_t group_id) const {
  write_string_attribute(group_id, "identifier", identifier);
  write_double_attribute(group_id, "diameter", diameter);
  write_double_attribute(group_id, "length", length);
  write_double_attribute(group_id, "ufrac", ufrac);
  write_double_attribute(group_id, "dfrac", dfrac);
}
void Pipe::load_from_hdf5(hid_t group_id) {
  // identifier = read_string(group_id, "identifier");
  length = read_double_attribute(group_id, "length");
  diameter = read_double_attribute(group_id, "diameter");
}

/*Pipe::Pipe(std::string identifier, Circuit* circuit, double diameter, double length, Node* unode, double ufrac, Node* dnode, double dfrac, double ficopt, double roughness, int ncell, double heat_input, double cfarea, int npar, double qcrit, double Kforward, int flowreg)
    : identifier(identifier), circuit(circuit), diameter(diameter), length(length), unode(unode), ufrac(ufrac), dnode(dnode), dfrac(dfrac), ncell(ncell), heat_input(heat_input), npar(npar), qcrit(qcrit), Kforward_old(Kforward), Kforward(Kforward), cfarea1(cfarea1) {

     double delx = length / ncell;
    double delz = (dnode.elevation - unode.elevation) / ncell;

    if (delz - delx > 1.E-6) {
        std::cout << "length is less than the elevation difference. stopping " << identifier << std::endl;
        exit(EXIT_FAILURE);
    }

    if (cfarea == NULL) {
        cfarea = M_PI * std::pow(diameter, 2) / 4.0;
    }

    cfarea *= npar;

    nodes.reserve(ncell - 1);
    for (int i = 0; i < ncell - 1; ++i) {
        Node node(identifier + "_node" + std::to_string(i), delx * cfarea, unode.elevation + (dnode.elevation - unode.elevation) * (i + 1) / ncell);
        node.i = i;
        nodes.push_back(node);
    }

    faces.reserve(ncell);
    for (int i = 0; i < ncell; ++i) {
        if (i == 0 && ncell == 1) {
            faces.push_back(PFace(i, *this, unode, ufrac, dnode, dfrac, diameter, cfarea, delx, delz, ficopt, roughness));
        } else if (i == 0) {
            faces.push_back(PFace(i, *this, unode, ufrac, nodes[0], NULL, diameter, cfarea, delx, delz, ficopt, roughness));
        } else if (i == ncell - 1) {
            faces.push_back(PFace(i, *this, nodes[ncell - 2], NULL, dnode, dfrac, diameter, cfarea, delx, delz, ficopt, roughness));
        } else {
            faces.push_back(PFace(i, *this, nodes[i - 1], NULL, nodes[i], NULL, diameter, cfarea, delx, delz, ficopt, roughness));
        }
    }

    for (int i = 0; i < ncell - 1; ++i) {
        nodes[i].ifaces.push_back(faces[i]);
        nodes[i].ofaces.push_back(faces[i + 1]);
    }

    unode.ofaces.push_back(faces[0]);

    dnode.ifaces.push_back(faces[ncell - 1]);

    mflow = 0.0;
 }*/

void Pipe::add_wall(double thk, std::string solname, std::string sollib, int restraint) {
  // for (auto& face : faces) {
    // face.wall = Wall(thk,solname,sollib,restraint);
  // }
}

// void Pipe::update_mflow() {
    // std::vector<double> mface;
    // for (auto& face : faces) {
        // mface.push_back(face.mflow);
    // }
    // if (!mface.empty()) {
        // mflow = std::accumulate(mface.begin(), mface.end(), 0.0) / mface.size();
    // }
// }



} // namespace opensd

File: opensd/pipe.py
import math
import numpy as np
import lxml.etree as ET
from opensd.face import PFace

class Pipe(object):

    def __init__(self,identifier,circuit,diameter,length,unode,ufrac,dnode,dfrac,fricopt,roughness,ncell,heat_input,cfarea,npar,qcrit,Kforward,flowreg):
        self.identifier=identifier
        self.circuit = circuit
        self.flowreg = flowreg

        self.diameter=diameter
        self.length=length

        self.ncell=ncell
        self.unode=unode
        self.ufrac = ufrac
        self.dnode=dnode
        self.dfrac = dfrac
        
        self.npar = npar
        self.qcrit = qcrit
        self.Kforward_old = self.Kforward = Kforward
        self.fricopt = fricopt

        delx = length/ncell
        delz = (dnode.elevation - unode.elevation)/ncell
        if delz - delx > 1.E-6:
            print ("length is less than the elevation difference. stopping",self.identifier)
            sys.exit()
        
        self.cfarea1 = cfarea
        if cfarea is None:
            cfarea=math.pi*diameter**2/4.
            
        self.cfarea = cfarea * npar
            
        self.heat_input = heat_input
        self.roughness = roughness

        # self.nodes = []
        # for i in range(ncell-1):
            # node=circuit.add_node(identifier+"_node"+str(i),volume=delx*cfarea,elevation=unode.elevation+(dnode.elevation-unode.elevation)*(i+1)/ncell)
            # node.i = i
            # self.nodes.append(node)
        
        # self.faces = []
        # for i in range(ncell):
            # if i==0 and ncell==1:
                # self.faces.append(PFace(i,self,unode,ufrac,dnode,dfrac,diameter,cfarea,delx,delz,fricopt,roughness))
            # elif i==0:
                # self.faces.append(PFace(i,self,unode,ufrac,self.nodes[0],None,diameter,cfarea,delx,delz,fricopt,roughness))
            # elif i==ncell-1:
                # self.faces.append(PFace(i,self,self.nodes[ncell-2],None,dnode,dfrac,diameter,cfarea,delx,delz,fricopt,roughness))
            # else:
                # self.faces.append(PFace(i,self,self.nodes[i-1],None,self.nodes[i],None,diameter,cfarea,delx,delz,fricopt,roughness))
            
        # for i in range(ncell-1):
            # self.nodes[i].ifaces.append(self.faces[i])
            # self.nodes[i].ofaces.append(self.faces[i+1])

        # unode.ofaces.append(self.faces[0])
        unode.volume = unode.volume + 0.5*delx*self.cfarea

        # dnode.ifaces.append(self.faces[-1])
        dnode.volume = dnode.volume + 0.5*delx*self.cfarea
        
        self.mflow = 0.
        
    def add_wall(self,thk,solname,sollib,restraint):
        wall = Wall(thk,solname,sollib,restraint)
        # for face in self.faces:
            # face.wall = wall

    def to_xml_element(self,element):
        subelement = ET.SubElement(element, "pipe")
        subelement.set("identifier", self.identifier)
        subelement.set("diameter",   str(self.diameter))
        subelement.set("length",     str(self.length))
        subelement.set("ncell",      str(self.ncell))
        subelement.set("unode",      str(self.unode.identifier))
        subelement.set("dnode",      str(self.dnode.identifier))
        subelement.set("roughness",  str(self.roughness))
        subelement.set("fricopt",    str(self.fricopt))
        subelement.set("cfarea",     str(self.cfarea))
        subelement.set("heat_input", str(self.heat_input))
        subelement.set("Kforward",   str(self.Kforward))
        subelement.set("ufrac",      "-1" if self.ufrac is None else str(self.ufrac))
        subelement.set("dfrac",      "-1" if self.dfrac is None else str(self.dfrac))

class Wall(object):
    def __init__(self,thk,solname,sollib,restraint):
        if sollib == "User":
            import sys
            import os
            sys.path.insert(0,os.getcwd() + "/")
            mod = __import__(solname)
            mat_clas = getattr(mod,'solid')
            self.mech_gues = mat_clas('mech')
        elif sollib == "thinmam":
            import thinmam
            self.mech_gues = thinmam.state(solname,'mech')
        else:
            print ("solid material not found. stopping",solname,sollib)
            sys.exit()
        if restraint == "long":
            self.c1 = 5./4. - self.mech_gues.poissons_ratio()
        else:
            print ("restraint option not found. stopping",self.identifier,restraint)
            sys.exit()
        self.thk = thk
